src/job_market.rs

```diff
@@ -0,0 +1,154 @@
+use parking_lot::{Condvar, Mutex};
+use std::{collections::VecDeque, sync::Arc};
+
+/// A market for synchronising the sharing of jobs.
+///
+/// Maintains synchronisation for multiple threads, including shutdown behaviour once one finishes
+/// or panics.
+pub struct JobMarket<Job> {
+    /// Get notified when there is a new job to handle.
+    has_new_job: Arc<Condvar>,
+    /// The market that we share.
+    market: Arc<Mutex<JobMarketInner<Job>>>,
+}
+
+impl<Job> Clone for JobMarket<Job> {
+    fn clone(&self) -> Self {
+        Self {
+            has_new_job: Arc::clone(&self.has_new_job),
+            market: Arc::clone(&self.market),
+        }
+    }
+}
```

Comment:
Idea for a future refactoring to make this code even more intuitive for new contributors...

Conceptually this is creating a new broker/recruiter rather than a new market. Perhaps a follow-up refactoring might rename `JobMarket` to `JobBroker` and `JobMarketInner` to `JobMarket` (or similar). Then we can describe behavior more intuitively: "the last broker closes the market" rather than "the last market closes the inner market" for instance.

Similarly `JobMarketInner::jobs` might be renamed `JobMarketInner::job_batches` since a `Job` in this implementation is defined as a single record, while there are currently some places where we have a `job: VecDequeue<Job>` (e.g. `if let Some(job) = market.jobs.pop()` would become `if let Some(job_batch) = market.job_batches.pop()` or similar).